heuristic/repair_operators.py

```diff
@@ -722,3 +747,192 @@ def worst_employee_regret_repair(competencies, t_covered_by_shift, employee_with
+
+        if sum(shifts.values()) == 0:
+            return repair_set
+
+def week_demand_based_repair_greedy(shifts_in_week, competencies, t_covered_by_shift,
+                             employee_with_competencies, employee_with_competency_combination,
+                             demand, time_step, time_periods_in_week, combined_time_periods_in_week,
+                             employees, contracted_hours, shifts_with_demand, 
+                             invalid_shifts, shift_combinations_violating_daily_rest,
+                             shift_sequences_violating_daily_rest,
+                             weeks, shifts_at_day, L_C_D, shifts_overlapping_t, state, destroy_set,
+                             week):
+
+    repair_set = []
+    employees_changed = employees
+    destroy_set = destroy_set.copy()
+    saturdays = [5 + j * 7 for j in week]
+    sundays = [6 + j * 7 for j in week]
+    shifts = {shift: demand for shift, demand in shifts_with_demand.items() if shift in shifts_in_week[week[0]]}
+    while(True):
+        below_minimum_demand(state, destroy_set, employee_with_competencies, demand, competencies, t_covered_by_shift)
+        delta_calculate_negative_deviation_from_contracted_hours(state, employees_changed, contracted_hours, weeks, time_periods_in_week, competencies, time_step)
+        calculate_deviation_from_demand(state, competencies, t_covered_by_shift, employee_with_competencies, demand, destroy_set)
+        
+        shift = max(shifts.items(), key=itemgetter(1))[0]
+        
+        possible_employees = [(e, score) for score, e in
+                              employee_with_competency_combination[(0,)] if (sum(
+                state.x[e, t, v] for t, v in shifts_at_day[int(shift[0] / 24)])) == 0]
+        #Husker ikke om jeg kjørte denne med week eller weeks
+        if(max([sum(state.soft_vars["deviation_contracted_hours"][e[0],j] for j in weeks) for e in possible_employees]) < shift[1]):
+            return repair_set
+        
+        
+        objective_values = {}
+        for e, score in possible_employees:
+            objective_values[e] = employee_shift_value(state, e, shift, saturdays, sundays, invalid_shifts, shift_combinations_violating_daily_rest, shift_sequences_violating_daily_rest, shifts_in_week, weeks, shifts_at_day, week[0], 0)
+
+        max_value = max(objective_values.items(), key=itemgetter(1))[1]
+        employee = choice([key for key, value in objective_values.items() if value == max_value])
+        repair_set.append(set_x(state, t_covered_by_shift, employee, shift[0], shift[1], 1, None))
+        employees_changed = [employee]
+        destroy_set = [(employee, shift[0], shift[1])]
+        shifts[shift] -= 1
+        if shifts[shift] == 0:
+            del shifts[shift]
+
+        if sum(shifts.values()) == 0:
+            return repair_set
+        
+
+def mip_week_operator_2(  employees, shifts_in_week, competencies, time_periods_in_week, combined_time_periods_in_week, 
```

Comment:
How is the runtime and results for this?

Reply:
Doesn't seem like its taking too much time. The operator that takes the longest at the moment is worst_employee_regret_repair